File: Ejercicios/Ejercicio2/1.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Evento 
{
public:
    string nombre;
    string fecha;
    vector<string> asistentes;

    Evento(string nom, string fec) : nombre(nom), fecha(fec) {}

    void agregarAsistente(const string& asistente)
     {
        asistentes.push_back(asistente);
        sort(asistentes.begin(), asistentes.end());
    }

    void mostrarAsistentes() const 
    {
        for (const auto& asistente : asistentes) 
        {
            cout << asistente << endl;
        }
    }
};

class Organizador 
{
private:
    vector<Evento> eventos;

public:
    void agregarEvento(const string& nombre, const string& fecha) 
    {
        auto it = find_if(eventos.begin(), eventos.end(),
        [&nombre](const Evento& evento) { return evento.nombre == nombre; });

        if (it != eventos.end()) {
            cout << "Evento con el nombre " << nombre << " ya existe." << endl;
            return;
        }

        eventos.push_back(Evento(nombre, fecha));
        cout << "Evento agregado exitosamente." << endl;
    }

    void agregarAsistente(const string& nombreEvento, const string& asistente) 
    {
        auto it = find_if(eventos.begin(), eventos.end(),
                          [&nombreEvento](const Evento& evento) { return evento.nombre == nombreEvento; });

        if (it != eventos.end()) 
        {
            it->agregarAsistente(asistente);
            cout << "Asistente agregado exitosamente." << endl;
        }
         else 
        {
            cout << "Evento no encontrado." << endl;
        }
    }

    void mostrarAsistentes(const string& nombreEvento) const 
    {
        auto it = find_if(eventos.begin(), eventos.end(),
                          [&nombreEvento](const Evento& evento) { return evento.nombre == nombreEvento; });

        if (it != eventos.end()) 
        {
            it->mostrarAsistentes();
        } 
        else 
        {
            cout << "Evento no encontrado." << endl;
        }
    }

    void buscarEventosPorFecha(const string& fecha) const 
    {
        bool encontrado = false;
        for (const auto& evento : eventos)
         {
            if (evento.fecha == fecha) 
            {
                cout << "Evento: " << evento.nombre << endl;
                encontrado = true;
            }
        }
        if (!encontrado) 
        {
            cout << "No se encontraron eventos en la fecha " << fecha << endl;
        }
    }
};

void mostrarMenu() 
{
    cout << "Menu de Gestion de Eventos" << endl;
    cout << "1. Agregar evento" << endl;
    cout << "2. Agregar asistente a evento" << endl;
    cout << "3. Mostrar asistentes de un evento" << endl;
    cout << "4. Buscar eventos por fecha" << endl;
    cout << "5. Salir" << endl;
}

int main() 
{
    Organizador organizador;
    int opcion;
    do 
    {
        mostrarMenu();
        cout << "Seleccione una opcion: ";
        cin >> opcion;

        string nombre, fecha, asistente;

        switch (opcion)
         {
            case 1:
                cout << "Ingrese nombre del evento: ";
                cin.ignore();
                getline(cin, nombre);
                cout << "Ingrese fecha del evento: ";
                getline(cin, fecha);
                organizador.agregarEvento(nombre, fecha);
                break;
            case 2:
                cout << "Ingrese nombre del evento: ";
                cin.ignore();
                getline(cin, nombre);
                cout << "Ingrese nombre del asistente: ";
                getline(cin, asistente);
                organizador.agregarAsistente(nombre, asistente);
                break;
            case 3:
                cout << "Ingrese nombre del evento: ";
                cin.ignore();
                getline(cin, nombre);
                organizador.mostrarAsistentes(nombre);
                break;
            case 4:
                cout << "Ingrese fecha del evento (YYYY-MM-DD): ";
                cin.ignore();
                getline(cin, fecha);
                organizador.buscarEventosPorFecha(fecha);
                break;
            case 5:
                cout << "Saliendo..." << endl;
                break;
            default:
                cout << "Opcion no valida. Intente nuevamente." << endl;
        }
    } while (opcion != 5);

    return 0;
}

File: Ejercicios/Ejercico1/1.java
import java.util.*;

class Producto {
    String codigo;
    String descripcion;
    int cantidad;

    public Producto(String cod, String desc, int cant) {
        codigo = cod;
        descripcion = desc;
        cantidad = cant;
    }
}

class Inventario {
    private Map<String, Producto> productos = new HashMap<>();

    public void agregarProducto(String codigo, String descripcion, int cantidad) {
        if (productos.containsKey(codigo)) {
            System.out.println("Producto con el codigo " + codigo + " ya existe.");
            return;
        }
        productos.put(codigo, new Producto(codigo, descripcion, cantidad));
        System.out.println("Producto agregado exitosamente.");
    }

    public void actualizarCantidad(String codigo, int cantidad) {
        if (productos.containsKey(codigo)) {
            productos.get(codigo).cantidad = cantidad;
            System.out.println("Cantidad actualizada exitosamente.");
        } else {
            System.out.println("Producto no encontrado.");
        }
    }

    public void buscarProducto(String codigo) {
        if (productos.containsKey(codigo)) {
            Producto prod = productos.get(codigo);
            System.out.println("Codigo: " + prod.codigo +
                               ", Descripcion: " + prod.descripcion +
                               ", Cantidad: " + prod.cantidad);
        } else {
            System.out.println("Producto no encontrado.");
        }
    }

    public void mostrarProductosOrdenados() {
        List<Producto> listaProductos = new ArrayList<>(productos.values());

        listaProductos.sort(Comparator.comparing(p -> p.descripcion));

        for (Producto prod : listaProductos) {
            System.out.println("Codigo: " + prod.codigo +
                               ", Descripcion: " + prod.descripcion +
                               ", Cantidad: " + prod.cantidad);
        }
    }
}

public class Main {
    public static void mostrarMenu() {
        System.out.println("Menu de Gestion de Inventario");
        System.out.println("1. Agregar producto");
        System.out.println("2. Actualizar cantidad");
        System.out.println("3. Buscar producto");
        System.out.println("4. Mostrar productos ordenados por descripcion");
        System.out.println("5. Salir");
    }

    public static void main(String[] args) {
        Inventario inventario = new Inventario();
        Scanner scanner = new Scanner(System.in);
        int opcion;

        do {
            mostrarMenu();
            System.out.print("Seleccione una opcion: ");
            opcion = scanner.nextInt();

            String codigo, descripcion;
            int cantidad;

            switch (opcion) {
                case 1:
                    System.out.print("Ingrese codigo: ");
                    codigo = scanner.next();
                    scanner.nextLine(); // consume newline
                    System.out.print("Ingrese descripcion: ");
                    descripcion = scanner.nextLine();
                    System.out.print("Ingrese cantidad: ");
                    cantidad = scanner.nextInt();
                    inventario.agregarProducto(codigo, descripcion, cantidad);
                    break;
                case 2:
                    System.out.print("Ingrese codigo: ");
                    codigo = scanner.next();
                    System.out.print("Ingrese nueva cantidad: ");
                    cantidad = scanner.nextInt();
                    inventario.actualizarCantidad(codigo, cantidad);
                    break;
                case 3:
                    System.out.print("Ingrese codigo: ");
                    codigo = scanner.next();
                    inventario.buscarProducto(codigo);
                    break;
                case 4:
                    inventario.mostrarProductosOrdenados();
                    break;
                case 5:
                    System.out.println("Saliendo...");
                    break;
                default:
                    System.out.println("Opcion no valida. Intente nuevamente.");
            }
        } while (opcion != 5);

        scanner.close();
    }
}
